File: webui/tools/executor.ts
/**
 * Tool Executor Engine — the heart of the agentic loop.
 *
 * Manages tool registration, concurrent execution, and orchestration.
 * Mirrors the original toolOrchestration.ts pattern:
 *   - Read-only tools run concurrently (max 10)
 *   - Write tools run serially
 */
import type { ToolDef, ToolContext, ToolResult, ToolCall, ToolCallResult, ToolProgress } from './types.js'
import { shellTool } from './shellTool.js'
import { fileReadTool, fileWriteTool, fileEditTool, globTool, grepTool } from './fileTools.js'
import { webFetchTool, webSearchTool } from './webTools.js'
import { todoWriteTool } from './todoTool.js'
import { enterPlanModeTool, exitPlanModeTool } from './planTool.js'
import { memoryWriteTool, memoryReadTool } from './memoryTool.js'
import { agentSpawnTool, agentStatusTool } from './agentTool.js'
import { notionSearchTool, notionReadPageTool, notionWritePageTool, notionDatabaseTool, notionBlockTool } from './notionTools.js'
import { databaseTool } from './databaseTool.js'
import { reportGenerateTool } from './reportTool.js'

// ─── Tool Registry ────────────────────────────────────────────

const toolRegistry = new Map<string, ToolDef>()

function registerTool(tool: ToolDef) {
  toolRegistry.set(tool.name, tool)
}

// Register all built-in tools
registerTool(shellTool)
registerTool(fileReadTool)
registerTool(fileWriteTool)
registerTool(fileEditTool)
registerTool(globTool)
registerTool(grepTool)
registerTool(webFetchTool)
registerTool(webSearchTool)
registerTool(todoWriteTool)
registerTool(enterPlanModeTool)
registerTool(exitPlanModeTool)
registerTool(memoryWriteTool)
registerTool(memoryReadTool)
registerTool(agentSpawnTool)
registerTool(agentStatusTool)
registerTool(notionSearchTool)
registerTool(notionReadPageTool)
registerTool(notionWritePageTool)
registerTool(notionDatabaseTool)
registerTool(notionBlockTool)
registerTool(databaseTool)
registerTool(reportGenerateTool)

export function getAllTools(): ToolDef[] {
  return Array.from(toolRegistry.values())
}

export function getTool(name: string): ToolDef | undefined {
  return toolRegistry.get(name)
}

// ─── Tool Schemas for AI ──────────────────────────────────────

/** Get all tool definitions in OpenAI function-calling format */
export function getToolSchemasForAI(): Array<{
  type: 'function'
  function: { name: string; description: string; parameters: Record<string, unknown> }
}> {
  return getAllTools().map((t) => ({
    type: 'function' as const,
    function: {
      name: t.name,
      description: t.description,
      parameters: t.inputSchema,
    },
  }))
}

/** Get all tool definitions in Anthropic format */
export function getToolSchemasForAnthropic(): Array<{
  name: string
  description: string
  input_schema: Record<string, unknown>
}> {
  return getAllTools().map((t) => ({
    name: t.name,
    description: t.description,
    input_schema: t.inputSchema,
  }))
}

// ─── Single Tool Execution ────────────────────────────────────

export async function executeTool(
  toolCall: ToolCall,
  ctx: ToolContext,
  onEvent?: (event: ToolExecutionEvent) => void,
): Promise<ToolCallResult> {
  const tool = getTool(toolCall.name)

  if (!tool) {
    return {
      tool_call_id: toolCall.id,
      name: toolCall.name,
      output: `Unknown tool: ${toolCall.name}. Available tools: ${getAllTools().map((t) => t.name).join(', ')}`,
      isError: true,
    }
  }

  onEvent?.({
    type: 'tool_start',
    toolId: toolCall.id,
    toolName: toolCall.name,
    input: toolCall.arguments,
  })

  // Permission check for high-risk tools
  if (tool.riskLevel === 'high' && ctx.requestPermission) {
    const description = `Execute ${tool.name}: ${JSON.stringify(toolCall.arguments).slice(0, 200)}`
    const allowed = await ctx.requestPermission(tool.name, toolCall.arguments, description)
    if (!allowed) {
      const result: ToolCallResult = {
        tool_call_id: toolCall.id,
        name: toolCall.name,
        output: 'Permission denied by user.',
        isError: true,
      }
      onEvent?.({ type: 'tool_end', toolId: toolCall.id, toolName: toolCall.name, result })
      return result
    }
  }

  try {
    // Execute with progress tracking
    const toolResult = await tool.execute(toolCall.arguments, {
      ...ctx,
      onProgress: (progress) => {
        progress.toolId = toolCall.id
        onEvent?.({ type: 'tool_progress', toolId: toolCall.id, toolName: toolCall.name, progress })
      },
    })

    const result: ToolCallResult = {
      tool_call_id: toolCall.id,
      name: toolCall.name,
      output: toolResult.output,
      isError: toolResult.isError,
    }

    onEvent?.({ type: 'tool_end', toolId: toolCall.id, toolName: toolCall.name, result })
    return result
  } catch (err: unknown) {
    const errMsg = err instanceof Error ? err.message : String(err)
    const result: ToolCallResult = {
      tool_call_id: toolCall.id,
      name: toolCall.name,
      output: `Tool execution error: ${errMsg}`,
      isError: true,
    }
    onEvent?.({ type: 'tool_end', toolId: toolCall.id, toolName: toolCall.name, result })
    return result
  }
}

// ─── Orchestrated Multi-Tool Execution ────────────────────────

/**
 * Execute multiple tool calls with smart concurrency:
 * - Read-only tools run in parallel (max concurrency: 10)
 * - Write tools run one at a time, serially
 */
export async function executeToolCalls(
  toolCalls: ToolCall[],
  ctx: ToolContext,
  onEvent?: (event: ToolExecutionEvent) => void,
): Promise<ToolCallResult[]> {
  const MAX_CONCURRENT = 10

  // Partition into read-only and write tools
  const readCalls: ToolCall[] = []
  const writeCalls: ToolCall[] = []

  for (const tc of toolCalls) {
    const tool = getTool(tc.name)
    if (tool?.isReadOnly) {
      readCalls.push(tc)
    } else {
      writeCalls.push(tc)
    }
  }

  const results: ToolCallResult[] = []

  // Run read-only tools concurrently (batched by MAX_CONCURRENT)
  for (let i = 0; i < readCalls.length; i += MAX_CONCURRENT) {
    const batch = readCalls.slice(i, i + MAX_CONCURRENT)
    const batchResults = await Promise.all(
      batch.map((tc) => executeTool(tc, ctx, onEvent))
    )
    results.push(...batchResults)
  }

  // Run write tools serially
  for (const tc of writeCalls) {
    const result = await executeTool(tc, ctx, onEvent)
    results.push(result)
  }

  return results
}

// ─── Event Types ──────────────────────────────────────────────

export type ToolExecutionEvent =
  | { type: 'tool_start'; toolId: string; toolName: string; input: Record<string, unknown> }
  | { type: 'tool_progress'; toolId: string; toolName: string; progress: ToolProgress }
  | { type: 'tool_end'; toolId: string; toolName: string; result: ToolCallResult }
  | { type: 'permission_request'; toolId: string; toolName: string; input: Record<string, unknown>; description: string }

File: api_server/tools/executor.py
"""Tool Executor Engine - manages tool registration, concurrent execution, and orchestration."""
import asyncio
from typing import Dict, List, Optional, Callable, Awaitable

from .types import (
    ToolDef,
    ToolContext,
    ToolCall,
    ToolCallResult,
    ToolProgress,
    ToolEvent,
    ToolStartEvent,
    ToolProgressEvent,
    ToolEndEvent,
)


tool_registry: Dict[str, ToolDef] = {}


def register_tool(tool: ToolDef) -> None:
    """Register a tool in the global registry."""
    tool_registry[tool.name] = tool


def get_tool(name: str) -> Optional[ToolDef]:
    """Get a tool definition by name."""
    return tool_registry.get(name)


def get_all_tools() -> List[ToolDef]:
    """Get all registered tools."""
    return list(tool_registry.values())


def get_tool_schemas_for_ai() -> List[Dict]:
    """Get all tool definitions in OpenAI function-calling format."""
    return [
        {
            "type": "function",
            "function": {
                "name": t.name,
                "description": t.description,
                "parameters": t.input_schema,
            },
        }
        for t in get_all_tools()
    ]


def get_tool_schemas_for_anthropic() -> List[Dict]:
    """Get all tool definitions in Anthropic format."""
    return [
        {
            "name": t.name,
            "description": t.description,
            "input_schema": t.input_schema,
        }
        for t in get_all_tools()
    ]


async def execute_tool(
    tool_call: ToolCall,
    ctx: ToolContext,
    on_event: Optional[Callable[[ToolEvent], Awaitable[None]]] = None,
) -> ToolCallResult:
    """Execute a single tool call with permission checking and event emission."""
    tool = get_tool(tool_call.name)

    if not tool:
        return ToolCallResult(
            tool_call_id=tool_call.id,
            name=tool_call.name,
            output=f"Unknown tool: {tool_call.name}. Available tools: {', '.join(t.name for t in get_all_tools())}",
            is_error=True,
        )

    if not hasattr(tool, 'execute') or tool.execute is None:
        return ToolCallResult(
            tool_call_id=tool_call.id,
            name=tool_call.name,
            output=f"Tool '{tool_call.name}' is registered but has no execute handler.",
            is_error=True,
        )

    if on_event:
        await on_event(
            ToolStartEvent(
                tool_id=tool_call.id,
                tool_name=tool_call.name,
                input=tool_call.arguments,
            )
        )

    risk_level = getattr(tool, 'risk_level', 'low')
    if risk_level == "high" and ctx.request_permission:
        desc = f"Execute {tool.name}: {str(tool_call.arguments)[:200]}"
        allowed = await ctx.request_permission(tool.name, tool_call.arguments, desc)
        if not allowed:
            result = ToolCallResult(
                tool_call_id=tool_call.id,
                name=tool_call.name,
                output="Permission denied by user.",
                is_error=True,
            )
            if on_event:
                await on_event(ToolEndEvent(tool_id=tool_call.id, tool_name=tool_call.name, result=result))
            return result

    try:
        async def progress_wrapper(progress: ToolProgress) -> None:
            progress.tool_id = tool_call.id
            await on_event(ToolProgressEvent(tool_id=tool_call.id, tool_name=tool_call.name, progress=progress)) if on_event else None

        tool_result = await tool.execute(tool_call.arguments, ToolContext(
            cwd=ctx.cwd,
            abort_signal=ctx.abort_signal,
            request_permission=ctx.request_permission,
            on_progress=progress_wrapper,
        ))

        result = ToolCallResult(
            tool_call_id=tool_call.id,
            name=tool_call.name,
            output=tool_result.output,
            is_error=tool_result.is_error,
        )

        await on_event(ToolEndEvent(tool_id=tool_call.id, tool_name=tool_call.name, result=result)) if on_event else None
        return result

    except Exception as err:
        err_msg = str(err)
        result = ToolCallResult(
            tool_call_id=tool_call.id,
            name=tool_call.name,
            output=f"Tool execution error: {err_msg}",
            is_error=True,
        )
        await on_event(ToolEndEvent(tool_id=tool_call.id, tool_name=tool_call.name, result=result)) if on_event else None
        return result


async def execute_tool_calls(
    tool_calls: List[ToolCall],
    ctx: ToolContext,
    on_event: Optional[Callable[[ToolEvent], Awaitable[None]]] = None,
) -> List[ToolCallResult]:
    """
    Execute multiple tool calls with smart concurrency:
    - Read-only tools run in parallel (max concurrency: 10)
    - Write tools run one at a time, serially
    """
    MAX_CONCURRENT = 10

    read_calls: List[ToolCall] = []
    write_calls: List[ToolCall] = []

    for tc in tool_calls:
        tool = get_tool(tc.name)
        if tool and tool.is_read_only:
            read_calls.append(tc)
        else:
            write_calls.append(tc)

    results: List[ToolCallResult] = []

    for i in range(0, len(read_calls), MAX_CONCURRENT):
        batch = read_calls[i : i + MAX_CONCURRENT]
        batch_results = await asyncio.gather(*[execute_tool(tc, ctx, on_event) for tc in batch])
        results.extend(batch_results)

    for tc in write_calls:
        result = await execute_tool(tc, ctx, on_event)
        results.append(result)

    return results